lib/QubitFlatfileImport.class.php

```diff
@@ -1917,32 +1914,77 @@ public function createContactInformation($actorId, $options)
   }
 
   /**
-   * Create a Qubit term or, if it already exists, fetch it
+   * Create or fetch Qubit terms from array, depending on if they already exist
+   * Must all be from same taxonomy.
    *
    * @param integer $taxonomyId  term taxonomy
-   * @param string $name  name of term
+   * @param mixed  $names Can be passed as single term name string
+   *                      or array of term names.
    * @param string $culture  culture code (defaulting to English)
    *
-   * @return object  created term or fetched object containing term data
+   * @return mixed created terms or fetched objects containing term data. Depending
+   *               on what was provided on input - string or array is returned.
    */
-  public static function createOrFetchTerm($taxonomyId, $name, $culture = 'en')
+  public static function createOrFetchTerm($taxonomyId, $names, $culture = 'en')
   {
-    $query = "SELECT t.id FROM term t LEFT JOIN term_i18n ti ON t.id=ti.id \r
-      WHERE t.taxonomy_id=? AND ti.name=? AND ti.culture=?";
+    if (!is_array($names))
+    {
+      $notArray = true;
+      $names = array($names);
+    }
 
-    $statement = QubitFlatfileImport::sqlQuery(
-      $query,
-      array($taxonomyId, $name, $culture)
-    );
+    // Retrieve terms in taxonomy from the term table once only for this taxonomy.
+    $query = "SELECT t.id, ti.name FROM term t LEFT JOIN term_i18n ti ON t.id=ti.id \r
+      WHERE t.taxonomy_id=? AND ti.culture=?";
 
-    $result = $statement->fetch(PDO::FETCH_OBJ);
-    if ($result)
+    $rows = QubitPdo::fetchAll($query, array($taxonomyId, $culture), array('fetchMode' => PDO::FETCH_ASSOC));
+
+    // Check if each term in array exists.
+    foreach($names as $name)
     {
-      return QubitTerm::getById($result->id);
+      if (null !== $key = QubitFlatfileImport::getTermIndex($rows, $name))
+      {
+        $terms[] = QubitTerm::getById($rows[$key]['id']);
+      }
+      else
+      {
+        $termsToCreate[] = $name;
+      }
     }
-    else
+
+    if (isset($termsToCreate))
     {
-      return QubitFlatfileImport::createTerm($taxonomyId, $name, $culture);
+      // Don't create duplicates
+      $termsToCreate = array_unique($termsToCreate);
+      foreach ($termsToCreate as $newterm)
+      {
+        $terms[] = QubitFlatfileImport::createTerm($taxonomyId, $newterm, $culture);
+      }
+    }
+
+    if (isset($notArray))
+    {
+      return $terms[0];
+    }
+    return $terms;
+  }
+
+  /**
+   * Retrieve term index for
+   * Must all be from same taxonomy.
+   *
+   * @param array  $rows  Results array from term query lookup
+   * @param string $name  Term name
+   *
+   * @return integer Index value if $name is found otherwise null
+   */
+  private static function getTermIndex($rows, $name)
+  {
+    foreach ($rows as $index => $row) {
```

Comment:
You could shorten it to this too I think:

`return array_search($name, array_column($rows, 'name'));`

Reply:
I tested the technique you are suggesting and it will not return the correct array key.  array_search will return the key of a _new_ array produced by array_column().  I need the key of the array returned by the PDO db lookup.

Reply:
D'oh... alright, never might that recommendation. ;)